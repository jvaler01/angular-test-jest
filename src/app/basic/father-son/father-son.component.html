<h1>Son</h1>
<hr>
<code class="mt-2">{{ client | json }}</code>
<hr>
<button data-test="btn-update" *ngIf="client" class="btn btn-outline-primary"(click)="onChange(2)">
    Change Id
</button>
<button data-test="btn-delete" *ngIf="client" class="btn btn-outline-danger ml-2" (click)="onDelete()">
    Remove client
</button>