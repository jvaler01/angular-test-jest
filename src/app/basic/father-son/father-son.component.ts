import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { Client } from '../interfaces';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.css']
})
export class FatherSonComponent implements OnInit {

  @Input()
  client?: Client;
  @Output() 
  onClientUpdated = new EventEmitter<Client>();
  @Output() 
  onClientDeleted = new EventEmitter();

  constructor() { }

  ngOnInit(): void {
  }

  onDelete(){
    this.client = undefined;
    this.onClientDeleted.emit();
  }

  onChange( id: number){
    if( !this.client ) return;
    this.client = {
      ...this.client,
      id
    }
    this.onClientUpdated.emit({ ...this.client });
  }
}
