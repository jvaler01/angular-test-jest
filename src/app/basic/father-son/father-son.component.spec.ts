import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from './father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ FatherSonComponent ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
    jest.clearAllMocks();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('check snapshoot', () => {
    expect( compiled ).toMatchSnapshot();
  });

  it('check buttons no client', () => {
    const buttons = compiled.querySelectorAll('button');
    expect( buttons.length ).toBe(0);
  });

  it('check buttons client', () => {
    component.client = { id: 1, name: "test"}
    fixture.detectChanges();
    const buttons = compiled.querySelectorAll('button');
    expect( buttons.length ).toBe(2);
  });

  it('check snapshoot client', () => {
    component.client = { id: 1, name: "test"}
    fixture.detectChanges();
    expect( compiled ).toMatchSnapshot();
  });

  it('check button emit deleted', () => {
    component.client = { id: 1, name: "test"}
    fixture.detectChanges();

    jest.spyOn( component.onClientDeleted, 'emit' );
    const button = compiled.querySelector('[data-test="btn-delete"]');
    button?.dispatchEvent( new Event('click') );
    expect( component.onClientDeleted.emit ).toHaveBeenCalled();
  });

  it('check button emit update', () => {
    component.client = { id: 1, name: "test"}
    fixture.detectChanges();

    jest.spyOn( component.onClientUpdated, 'emit' );
    const button = compiled.querySelector('[data-test="btn-update"]');
    button?.dispatchEvent( new Event('click') );
    expect( component.onClientUpdated.emit ).toHaveBeenCalledWith({
      id: 2,
      name: 'test'
    });
  });

  it('check button emit update if id != undefined and client != undefined', () => {
    jest.spyOn( component.onClientUpdated, 'emit' );
    component.onChange(10);
    expect( component.onClientUpdated.emit ).not.toHaveBeenCalled();
    
    component.client = { id: 1, name: "test"}
    fixture.detectChanges();
    component.onChange(10);

    expect( component.onClientUpdated.emit ).toHaveBeenCalledWith({
      id: 10,
      name: 'test'
    });
  });
});
