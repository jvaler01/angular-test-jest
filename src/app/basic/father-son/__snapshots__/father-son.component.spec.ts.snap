// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent check snapshoot 1`] = `
<div
  id="root1"
>
  <h1>
    Son
  </h1>
  <hr />
  <code
    class="mt-2"
  >
    
  </code>
  <hr />
  
  
</div>
`;

exports[`FatherSonComponent check snapshoot client 1`] = `
<div
  id="root4"
>
  <h1>
    Son
  </h1>
  <hr />
  <code
    class="mt-2"
  >
    {
  "id": 1,
  "name": "test"
}
  </code>
  <hr />
  <button
    class="btn btn-outline-primary"
    data-test="btn-update"
  >
     Change Id

  </button>
  
  <button
    class="btn btn-outline-danger ml-2"
    data-test="btn-delete"
  >
     Remove client

  </button>
  
</div>
`;
